dlc-oracle/src/main/scala/org/bitcoins/dlc/oracle/OracleEvent.scala

```diff
@@ -0,0 +1,268 @@
+    eventDescriptorTLV: RangeEventDescriptorV0TLV)
+    extends PendingOracleEvent
+    with RangeV0OracleEvent
+
+case class CompletedRangeV0OracleEvent(
+    pubkey: SchnorrPublicKey,
+    eventName: String,
+    signingVersion: SigningVersion,
+    maturationTime: Instant,
+    announcementSignature: SchnorrDigitalSignature,
+    eventDescriptorTLV: RangeEventDescriptorV0TLV,
+    attestation: FieldElement)
+    extends CompletedOracleEvent
+    with RangeV0OracleEvent {
+  override def attestations: Vector[FieldElement] = Vector(attestation)
+
+}
+
+sealed trait ExternalV0OracleEvent extends OracleEvent {
+  override def eventDescriptorTLV: ExternalEventDescriptorV0TLV
+  override def nonces: Vector[SchnorrNonce] = Vector(eventDescriptorTLV.nonce)
+}
+
+case class PendingExternalV0OracleEvent(
+    pubkey: SchnorrPublicKey,
+    eventName: String,
+    signingVersion: SigningVersion,
+    maturationTime: Instant,
+    announcementSignature: SchnorrDigitalSignature,
+    eventDescriptorTLV: ExternalEventDescriptorV0TLV)
+    extends PendingOracleEvent
+    with ExternalV0OracleEvent
+
+case class CompletedExternalV0OracleEvent(
+    pubkey: SchnorrPublicKey,
+    eventName: String,
+    signingVersion: SigningVersion,
+    maturationTime: Instant,
+    announcementSignature: SchnorrDigitalSignature,
+    eventDescriptorTLV: ExternalEventDescriptorV0TLV,
+    attestation: FieldElement)
+    extends CompletedOracleEvent
+    with ExternalV0OracleEvent {
+  override def attestations: Vector[FieldElement] = Vector(attestation)
+}
+
+sealed trait LargeRangeV0OracleEvent extends OracleEvent {
+  override def eventDescriptorTLV: LargeRangeEventDescriptorV0TLV
+
+  override def nonces: Vector[SchnorrNonce] = eventDescriptorTLV.nonces
+}
+
+case class PendingLargeRangeV0OracleEvent(
+    pubkey: SchnorrPublicKey,
+    eventName: String,
+    signingVersion: SigningVersion,
+    maturationTime: Instant,
+    announcementSignature: SchnorrDigitalSignature,
+    eventDescriptorTLV: LargeRangeEventDescriptorV0TLV)
+    extends PendingOracleEvent
+    with LargeRangeV0OracleEvent
+
+case class CompletedLargeRangeV0OracleEvent(
+    pubkey: SchnorrPublicKey,
+    eventName: String,
+    signingVersion: SigningVersion,
+    maturationTime: Instant,
+    announcementSignature: SchnorrDigitalSignature,
+    eventDescriptorTLV: LargeRangeEventDescriptorV0TLV,
+    attestations: Vector[FieldElement])
+    extends CompletedOracleEvent
+    with LargeRangeV0OracleEvent
+
+object OracleEvent {
+
+  def apply(eventDbs: Vector[EventDb]): OracleEvent = {
+    val eventDb = eventDbs.head
+    require(eventDbs.forall(_.eventDescriptorTLV == eventDb.eventDescriptorTLV),
+            "EventDbs must all refer to the same event")
+
+    (eventDb.eventDescriptorTLV, eventDb.attestationOpt) match {
+      case (enum: EnumEventDescriptorV0TLV, Some(sig)) =>
+        require(eventDbs.size == 1, "Enum events may only have have eventDb")
```

Comment:
Maybe with the `SingleNonceEvent` stuff above this can be enforced at the type level instead of needing invariants?

Reply:
since `eventDbs` is a param we would still need these